zhaquirks/tuya/ts601_door.py

```diff
@@ -0,0 +1,230 @@
+"""Device handler for Tuya Zigbee Vibration Door Sensor.
+
+The device reports:
+- Vibration
+- Open/Closed door/window
+- Remaining battery percentage
+
+Extra details: https://www.aliexpress.com/item/1005004443361928.html?spm=a2g0o.order_list.order_list_main.53.35a81802wiZASs
+"""
+
+from zigpy.endpoint import Endpoint
+from zigpy.profiles import zha
+from zigpy.quirks import CustomCluster, CustomDevice
+from zigpy.zcl.clusters.general import (
+    Basic,
+    Groups,
+    Ota,
+    PowerConfiguration,
+    Scenes,
+    Time,
+)
+from zigpy.zcl.clusters.security import IasZone, ZoneStatus, ZoneType
+
+from zhaquirks.const import (
+    DEVICE_TYPE,
+    ENDPOINTS,
+    INPUT_CLUSTERS,
+    MODELS_INFO,
+    OUTPUT_CLUSTERS,
+    PROFILE_ID,
+    ZONE_STATE,
+)
+from zhaquirks.tuya import (
+    TUYA_CLUSTER_ID,
+    DPToAttributeMapping,
+    TuyaLocalCluster,
+    TuyaNewManufCluster,
+)
+
+DOOR_HANDLE_DP_ID = 1
+BATTERY_STATE_DP_ID = 3
+VIBRATION_DP_ID = 10
+
+DP_HANDLER_EP_ID = 1
+DOOR_HANDLE_EP_ID = 2
+VIBRATION_EP_ID = 3
+
+
+class CustomBasicCluster(CustomCluster, Basic):
+    """Custom Basic cluster to report power source."""
+
+    _CONSTANT_ATTRIBUTES = {
+        Basic.AttributeDefs.power_source.id: Basic.PowerSource.Battery
+    }
+
+
+class CustomTuyaPowerConfigurationCluster(TuyaLocalCluster, PowerConfiguration):
+    """Custom Power Configuration cluster that represents battery reports."""
+
+    _CONSTANT_ATTRIBUTES = {
+        PowerConfiguration.AttributeDefs.battery_size.id: PowerConfiguration.BatterySize.AAA,
+        PowerConfiguration.AttributeDefs.battery_quantity.id: 2,
+    }
+
+    def __init__(self, endpoint: Endpoint, is_server: bool = True):
+        """Init constructor that initializes battery attributes."""
+        super().__init__(endpoint=endpoint, is_server=is_server)
+
+        # Initialize remaining battery to 100% until a proper measurement arrives.
+        # Device measures battery in 1% steps, while the ZCL specifies it in 0.5% steps.
+        self.update_attribute(
+            attr_name=PowerConfiguration.AttributeDefs.battery_percentage_remaining.name,
+            value=100 * 2,
+        )
```

Comment:
Since this is done in `__init__`, this will update the attribute to 100% if HA is restarted.
Something like `if attr_id not in self._attr_cache` could be added before, but I think I'd just remove this for now.
HA should still initialize the sensor. It should just show something like "unknown". How long does it take for the device to report a value?

Reply:
Removed for now.

Reply:
As far as I've noticed, the device reports battery state when there's a change in percentage. This doesn't happen too often: either when there's some variability in the measured voltage (e.g. moves from 80% to 81% then back to 80%), or when there's an actual drop. Since this is a door sensor and is supposed to be used frequently, the value will probably get initialized fairly soon.